### PR TITLE
[#169971129] Integrate a role-based access control mechanism

### PR DESCRIPTION
**Pull request by alexgpeppe**
This PR aims to add a role-based access control mechanism.

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * ⚙️ [#169971129](https://www.pivotaltracker.com/story/show/169971129): Considerare l'integrazione di un meccanismo di autorizzazione di tipo RBAC


New dependencies added: <a href='https://www.npmjs.com/package/accesscontrol'><code>accesscontrol</code></a>.


<h2><a href="https://github.com/onury/accesscontrol#readme">accesscontrol</a></h2>
<p>Author: Onur Yildirim</p>
<p>Description: Role and Attribute based Access Control for Node.js</p>
<p>Homepage: <a href="https://github.com/onury/accesscontrol#readme">https://github.com/onury/accesscontrol#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>about 3 years ago</td></tr><tr><td>Last Updated</td><td>over 1 year ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>10</td></tr><tr><td>Direct Dependencies</td><td><a href='https://www.npmjs.com/package/notation'><code>notation</code></a></td></tr><tr><td>Keywords</td><td>access, access-control, acl, role, attribute, grant, deny, allow, reject, permission, action, possession, rbac, abac, crud, create, read, update, delete, resource, express, admin, user, glob, wildcard, policy, scope and context</td></tr>
</table>
This README is too long to show.

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 63fdac83267de3c0f494c92863eff26e56489ccb
</p>

File: src/controllers/organizationController.ts [hunk 1]
```diff
@@ -87,7 +88,10 @@ export default class OrganizationController {
     | IResponseSuccessRedirectToResource<Organization, Organization>
   > {
     return withUserFromRequest(req, async user => {
-      if (user.role !== UserRoleEnum.ORG_DELEGATE) {
+      const userPermission = accessControl
```
Comment (gunzip) [inline src/controllers/organizationController.ts]:
```suggestion
      const createOwnOrganization = accessControl
```

Comment (gunzip) [reply src/controllers/organizationController.ts]: or, better, extract and reuse:

`const userPerm = accessControl.can(user.role);`

File: src/controllers/organizationController.ts [hunk 2]
```diff
@@ -87,7 +92,9 @@ export default class OrganizationController {
     | IResponseSuccessRedirectToResource<Organization, Organization>
   > {
     return withUserFromRequest(req, async user => {
-      if (user.role !== UserRoleEnum.ORG_DELEGATE) {
+      const userPermissions = accessControl.can(user.role);
+      const permission = userPermissions.createOwn(ORGANIZATION_RESOURCE);
```
Comment (gunzip) [inline src/controllers/organizationController.ts]:
```suggestion
      const createOwnOrganization = userPermissions.createOwn(ORGANIZATION_RESOURCE);
```

File: src/acl/acl.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
+import { AccessControl } from "accesscontrol";
+import { UserRoleEnum } from "../generated/UserRole";
+
+// Resource constants
+export const ADMINISTRATION_RESOURCE = "administration";
```
Comment (gunzip) [inline src/acl/acl.ts]: what's the difference between administration and organization ?

Comment (alexgpeppe) [reply src/acl/acl.ts]: - `administration` is a public administration from IndicePA, ie. an administration that never started a registration process
- `organization` is a public administration from IO, ie. an administration whose registration process already started

[the remark this answers is not in the document]
Comment (alexgpeppe) [reply src/acl/acl.ts]: The user role enum is defined in the specs file and its typescript definition is generated by `io-utils` tool. The suffix is added by the tool in order to have different names for the enum definition (`UserRoleEnum`) and its type and codec (`UserRole`), so it's not possible to change it. I can rename the import, but still the original name would have the prefix; should I?

File: src/controllers/organizationController.ts [hunk 4]
```diff
@@ -173,46 +176,41 @@ export default class OrganizationController {
...
+            .map<
+              | IResponseErrorInternal
+              | IResponseSuccessJson<OrganizationCollection>
+            >(organization =>
+              ResponseSuccessJson({
+                items: [organization]
+              })
+            )
+            .getOrElse(
+              user.role === UserRoleEnum.ORG_MANAGER
```
Comment (gunzip) [inline src/controllers/organizationController.ts]: may we have a comment here explaining this condition ?

File: src/acl/acl.ts (hunk 3 above)
Comment (gunzip) [reply src/acl/acl.ts]: can we add a comment ?

[the remark this answers is not in the document]
Comment (gunzip) [reply src/acl/acl.ts]: no, it's ok

**Pull request merged**